### PR TITLE
Make WebPermissionController independent of the WebPage

### PR DESCRIPTION
**Pull request by RupinMittal**
#### 2a9bcbc919a9fa687506d51541efb66a80ee52f2
<pre>
Make WebPermissionController independent of the WebPage
<a href="https://bugs.webkit.org/show_bug.cgi?id=243597">https://bugs.webkit.org/show_bug.cgi?id=243597</a>

Reviewed by Chris Dumez and Sihui Liu.

Currently, the Permissions API is not supported for workers. In order for it
to be supported, the Permissions API should function without the need for a
page. Previously, WebPermissionController was created and owned by WebPage.
This patch makes PermissionController a singleton that is created by the
WebProcess. So now, instead of Permissions::query() using the page to get the
WebPermissionController and send an IPC message to WebPageProxy to obtain
permissions states, it can access the WebPermissionController singleton to
send an IPC message that will eventually call the correct WebPageProxy. Note
that a page is still required to obtain the permission state, but this
refactoring is the first step in supporting the Permissions API for workers.

* Source/WebCore/Modules/permissions/PermissionController.cpp: Copied from Source/WebCore/Modules/permissions/Permissions.h.
(WebCore::sharedController):
(WebCore::PermissionController::shared):
(WebCore::PermissionController::setSharedController):
* Source/WebCore/Modules/permissions/PermissionController.h:
* Source/WebCore/Modules/permissions/PermissionStatus.cpp:
(WebCore::PermissionStatus::PermissionStatus):
(WebCore::PermissionStatus::~PermissionStatus):
* Source/WebCore/Modules/permissions/PermissionStatus.h:
* Source/WebCore/Modules/permissions/Permissions.cpp:
(WebCore::Permissions::Permissions):
(WebCore::Permissions::query):
* Source/WebCore/Modules/permissions/Permissions.h:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::permissionController): Deleted.
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/ScriptExecutionContext.cpp:
(WebCore::ScriptExecutionContext::permissionController): Deleted.
* Source/WebCore/dom/ScriptExecutionContext.h:
* Source/WebCore/loader/EmptyClients.cpp:
(WebCore::pageConfigurationWithEmptyClients):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::permissionController): Deleted.
* Source/WebCore/page/Page.h:
* Source/WebCore/page/PageConfiguration.cpp:
(WebCore::PageConfiguration::PageConfiguration):
* Source/WebCore/page/PageConfiguration.h:
* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/WebPermissionControllerProxy.cpp: Copied from Source/WebCore/Modules/permissions/Permissions.h.
(WebKit::WebPermissionControllerProxy::WebPermissionControllerProxy):
(WebKit::WebPermissionControllerProxy::~WebPermissionControllerProxy):
(WebKit::WebPermissionControllerProxy::query):
* Source/WebKit/UIProcess/WebPermissionControllerProxy.h: Copied from Source/WebCore/Modules/permissions/Permissions.h.
* Source/WebKit/UIProcess/WebPermissionControllerProxy.messages.in: Copied from Source/WebCore/Modules/permissions/Permissions.h.
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::m_webPermissionController):
(WebKit::m_webLockRegistry): Deleted.
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp:
(WebKit::WebPermissionController::create):
(WebKit::WebPermissionController::query):
(WebKit::WebPermissionController::tryProcessingRequests):
(WebKit::WebPermissionController::WebPermissionController): Deleted.
* Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_appHighlightsVisible):
* Source/WebKit/WebProcess/WebProcess.cpp:
* Source/WebKitLegacy/mac/WebView/WebView.mm:
(-[WebView _commonInitializationWithFrameName:groupName:]):
(-[WebView initSimpleHTMLDocumentWithStyle:frame:preferences:groupName:]):
* Source/WebKitLegacy/win/WebView.cpp:
(WebView::initWithFrame):

Canonical link: <a href="https://commits.webkit.org/253216@main">https://commits.webkit.org/253216@main</a>
</pre>

### CODE REVIEW

File: Source/WebCore/Modules/permissions/PermissionController.h [hunk 1]
```diff
@@ -25,6 +25,7 @@
 
 #pragma once
 
+#include "Page.h"
```
Comment (cdumez) [inline Source/WebCore/Modules/permissions/PermissionController.h]: Please use a Forward declaration instead of an include.

File: Source/WebKit/UIProcess/WebPageProxy.cpp [hunk 2]
```diff
@@ -8840,7 +8840,7 @@ void WebPageProxy::revokeGeolocationAuthorizationToken(const String& authorizati
     m_geolocationPermissionRequestManager.revokeAuthorizationToken(authorizationToken);
 }
 
-void WebPageProxy::queryPermission(const ClientOrigin& clientOrigin, const PermissionDescriptor& descriptor, CompletionHandler<void(std::optional<PermissionState>, bool shouldCache)>&& completionHandler)
+void WebPageProxy::queryPermission(const ClientOrigin& clientOrigin, const PermissionDescriptor& descriptor, CompletionHandler<void(std::optional<PermissionState>, bool shouldCache)>& completionHandler)
```
Comment (cdumez) [inline Source/WebKit/UIProcess/WebPageProxy.cpp]: Why this change? This doesn't look right.

File: Source/WebKit/UIProcess/WebPermissionControllerProxy.cpp [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    m_process.removeMessageReceiver(Messages::WebPermissionControllerProxy::messageReceiverName());
+}
+
+void WebPermissionControllerProxy::query(const WebCore::ClientOrigin& clientOrigin, const WebCore::PermissionDescriptor& descriptor, WebPageProxyIdentifier identifier, CompletionHandler<void(std::optional<WebCore::PermissionState>, bool shouldCache)>&& completionHandler)
+{
+    auto webPageProxy = m_process.webPage(identifier);
+    if (!webPageProxy)
+        completionHandler(WebCore::PermissionState::Prompt, false);
+
+    webPageProxy->queryPermission(clientOrigin, descriptor, completionHandler);
```
Comment (cdumez) [inline Source/WebKit/UIProcess/WebPermissionControllerProxy.cpp]: WTFMove(completionHandler)

File: Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp [hunk 4]
```diff
@@ -26,33 +26,33 @@
...
+#include <WebCore/Page.h>
 #include <WebCore/PermissionObserver.h>
 
 namespace WebKit {
 
-Ref<WebPermissionController> WebPermissionController::create(WebPage& page)
+Ref<WebPermissionController> WebPermissionController::create()
 {
-    return adoptRef(*new WebPermissionController(page));
+    return adoptRef(*new WebPermissionController());
```
Comment (cdumez) [inline Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp]: No need for `()`:
`return adoptRef(*new WebPermissionController);`

File: Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp [hunk 5]
```diff
@@ -26,33 +26,33 @@
...
 }
 
-WebPermissionController::WebPermissionController(WebPage& page)
-    : m_page(page)
+WebPermissionController::WebPermissionController()
```
Comment (cdumez) [inline Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp]: WebPermissionController::WebPermissionController() = default;

File: Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp [hunk 6]
```diff
@@ -26,33 +26,33 @@
...
-    if (!m_page)
-        return completionHandler({ });
-
     auto cachedResult = queryCache(origin, descriptor);
     if (cachedResult != WebCore::PermissionState::Prompt)
         return completionHandler(cachedResult);
 
-    m_requests.append(PermissionRequest { WTFMove(origin), WTFMove(descriptor), WTFMove(completionHandler) });
+    auto webPageProxyIdentifier = WebPage::fromCorePage(page).webPageProxyIdentifier();
+    m_requests.append(PermissionRequest { WTFMove(origin), WTFMove(descriptor), WTFMove(completionHandler), WTFMove(webPageProxyIdentifier) });
```
Comment (cdumez) [inline Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp]: I think it'd look nicer if the completion handler came last.

Comment (cdumez) [reply Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp]: No WTFMove() for webPageProxyIdentifier since this is a uint64_t.

File: Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.h [hunk 7]
```diff
@@ -33,17 +34,15 @@
...
-class WebPage;
-
 class WebPermissionController final : public CanMakeWeakPtr<WebPermissionController>, public WebCore::PermissionController {
 public:
-    static Ref<WebPermissionController> create(WebPage&);
+    static Ref<WebPermissionController> create();
 
 private:
-    explicit WebPermissionController(WebPage&);
+    explicit WebPermissionController();
```
Comment (cdumez) [inline Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.h]: You can drop the explicit.

File: Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.h [hunk 8]
```diff
@@ -62,6 +60,7 @@ class WebPermissionController final : public CanMakeWeakPtr<WebPermissionControl
         WebCore::ClientOrigin origin;
         WebCore::PermissionDescriptor descriptor;
         CompletionHandler<void(std::optional<WebCore::PermissionState>)> completionHandler;
+        WebPageProxyIdentifier identifier;
```
Comment (cdumez) [inline Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.h]: I think this should come before the completion handler.

File: Tools/MiniBrowser/MiniBrowser.xcodeproj/project.pbxproj [hunk 9]
```diff
@@ -3,7 +3,7 @@
 	archiveVersion = 1;
 	classes = {
 	};
-	objectVersion = 46;
+	objectVersion = 52;
```
Comment (cdumez) [inline Tools/MiniBrowser/MiniBrowser.xcodeproj/project.pbxproj]: Please revert this change.

**Review by cdumez — CHANGES_REQUESTED** [review]

File: Source/WebKit/UIProcess/WebPageProxy.cpp (hunk 2 above)
Comment (RupinMittal) [reply Source/WebKit/UIProcess/WebPageProxy.cpp]: I did this because WebPermissionControllerProxy takes in the completionHandler as an rvalue reference, so it when it passes it to WebPageProxy, WebPageProxy should take it in as a lvalue reference right?

Comment (cdumez) [reply Source/WebKit/UIProcess/WebPageProxy.cpp]: You did this because you made a mistake and forgot a WTFMove(completionHandler), like I commented below.

File: Tools/MiniBrowser/MiniBrowser.xcodeproj/project.pbxproj (hunk 9 above)
Comment (RupinMittal) [reply Tools/MiniBrowser/MiniBrowser.xcodeproj/project.pbxproj]: I'm curious, why did this change happen in the first place? (I didn't change it manually)

Comment (cdumez) [reply Tools/MiniBrowser/MiniBrowser.xcodeproj/project.pbxproj]: I don't know, Xcode...

File: Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp [hunk 10]
```diff
@@ -104,11 +104,10 @@ void WebPermissionController::tryProcessingRequests()
         }
 
         currentRequest.isWaitingForReply = true;
-        m_page->sendWithAsyncReply(Messages::WebPageProxy::QueryPermission(currentRequest.origin, currentRequest.descriptor), [this, weakThis = WeakPtr { *this }](auto state, bool shouldCache) {
+        WebProcess::singleton().sendWithAsyncReply(Messages::WebPermissionControllerProxy::Query(currentRequest.origin, currentRequest.descriptor, currentRequest.identifier), [this, weakThis = WeakPtr { *this }](auto state, bool shouldCache) {
```
Comment (szewai) [inline Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp]: Before sending message to UI process, you can check if page exists with:
`WebProcess::singleton().webPage(currentRequest.identifier)`, if not you may just invoke completionHandler of the request

File: Source/WebKit/UIProcess/WebPermissionControllerProxy.cpp [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+WebPermissionControllerProxy::~WebPermissionControllerProxy()
+{
+    m_process.removeMessageReceiver(Messages::WebPermissionControllerProxy::messageReceiverName());
+}
+
+void WebPermissionControllerProxy::query(const WebCore::ClientOrigin& clientOrigin, const WebCore::PermissionDescriptor& descriptor, WebPageProxyIdentifier identifier, CompletionHandler<void(std::optional<WebCore::PermissionState>, bool shouldCache)>&& completionHandler)
+{
+    auto webPageProxy = m_process.webPage(identifier);
+    if (!webPageProxy)
+        completionHandler(WebCore::PermissionState::Prompt, false);
```
Comment (szewai) [inline Source/WebKit/UIProcess/WebPermissionControllerProxy.cpp:52]: You should return here, otherwise, the next line will be executed

File: Source/WebKit/UIProcess/WebPermissionControllerProxy.cpp [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+#include "config.h"
+#include "WebPermissionControllerProxy.h"
+
+#include "Connection.h"
+#include "WebPageProxyIdentifier.h"
+#include "WebPermissionControllerProxyMessages.h"
+#include "WebProcessProxy.h"
+#include <WebCore/ClientOrigin.h>
+#include <WebCore/PermissionDescriptor.h>
+#include <WebCore/PermissionState.h>
```
Comment (szewai) [inline Source/WebKit/UIProcess/WebPermissionControllerProxy.cpp]: No need to include same headers in both .h and .cpp, and please use forward declaration if possible

File: Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp [hunk 13]
```diff
@@ -26,33 +26,33 @@
 #include "config.h"
 #include "WebPermissionController.h"
 
-
 #include "WebPage.h"
+#include "WebPageProxyIdentifier.h"
```
Comment (szewai) [inline Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp:30]: No need to include the same header in .h and .cpp

File: Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp [hunk 14]
```diff
@@ -88,7 +88,7 @@ void WebPermissionController::updateCache(const WebCore::ClientOrigin& origin, c
 
 void WebPermissionController::tryProcessingRequests()
 {
-    if (m_requests.isEmpty() || !m_page)
+    if (m_requests.isEmpty())
```
Comment (szewai) [inline Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp]: You can remove this, the while loop won't run if `m_requests` is empty

File: Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp [hunk 15]
```diff
@@ -104,11 +104,10 @@ void WebPermissionController::tryProcessingRequests()
         }
 
         currentRequest.isWaitingForReply = true;
-        m_page->sendWithAsyncReply(Messages::WebPageProxy::QueryPermission(currentRequest.origin, currentRequest.descriptor), [this, weakThis = WeakPtr { *this }](auto state, bool shouldCache) {
+        WebProcess::singleton().sendWithAsyncReply(Messages::WebPermissionControllerProxy::Query(currentRequest.origin, currentRequest.descriptor, currentRequest.identifier), [this, weakThis = WeakPtr { *this }](auto state, bool shouldCache) {
             RefPtr protectedThis { weakThis.get() };
             if (!protectedThis)
                 return;
-
```
Comment (szewai) [inline Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp]: Nit: no need to remove the blank line.

**Review by szewai — CHANGES_REQUESTED** [review]

File: Source/WebCore/Modules/permissions/PermissionController.h (hunk 1 above)
Comment (RupinMittal) [reply Source/WebCore/Modules/permissions/PermissionController.h]: Fixed!

File: Source/WebKit/UIProcess/WebPageProxy.cpp (hunk 2 above)
Comment (RupinMittal) [reply Source/WebKit/UIProcess/WebPageProxy.cpp]: Ah you're right--I see it. Fixed!

File: Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp (hunk 4 above)
Comment (RupinMittal) [reply Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp]: Fixed!

File: Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp (hunk 5 above)
Comment (RupinMittal) [reply Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp]: Fixed!

File: Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp (hunk 6 above)
Comment (RupinMittal) [reply Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp]: Fixed!

File: Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.h (hunk 8 above)
Comment (RupinMittal) [reply Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.h]: Fixed!

File: Tools/MiniBrowser/MiniBrowser.xcodeproj/project.pbxproj (hunk 9 above)
Comment (RupinMittal) [reply Tools/MiniBrowser/MiniBrowser.xcodeproj/project.pbxproj]: Fixed!

File: Source/WebKit/UIProcess/WebPermissionControllerProxy.cpp (hunk 11 above)
Comment (RupinMittal) [reply Source/WebKit/UIProcess/WebPermissionControllerProxy.cpp:52]: Fixed!

File: Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp (hunk 10 above)
Comment (RupinMittal) [reply Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp]: Fixed!

File: Source/WebKit/UIProcess/WebPermissionControllerProxy.cpp (hunk 12 above)
Comment (RupinMittal) [reply Source/WebKit/UIProcess/WebPermissionControllerProxy.cpp]: Fixed!

File: Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp (hunk 13 above)
Comment (RupinMittal) [reply Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp:30]: Fixed!

File: Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp (hunk 14 above)
Comment (RupinMittal) [reply Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp]: Fixed!

File: Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp (hunk 15 above)
Comment (RupinMittal) [reply Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp]: Fixed!

File: Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.h (hunk 7 above)
Comment (RupinMittal) [reply Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.h]: Fixed!

File: Source/WebKit/UIProcess/WebPermissionControllerProxy.cpp (hunk 3 above)
Comment (RupinMittal) [reply Source/WebKit/UIProcess/WebPermissionControllerProxy.cpp]: Fixed!

File: Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp [hunk 16]
```diff
@@ -26,33 +26,32 @@
...
 {
-    if (!m_page)
-        return completionHandler({ });
-
     auto cachedResult = queryCache(origin, descriptor);
     if (cachedResult != WebCore::PermissionState::Prompt)
         return completionHandler(cachedResult);
 
-    m_requests.append(PermissionRequest { WTFMove(origin), WTFMove(descriptor), WTFMove(completionHandler) });
+    auto pageIdentifier = WebPage::fromCorePage(page).identifier();
```
Comment (cdumez) [inline Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp:53]: Why are you storing a PageIdentifier when you really need a WebPageProxyIdentifier?

We would probably call webPageProxyIdentifier() here.

File: Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp [hunk 17]
```diff
@@ -103,8 +99,14 @@ void WebPermissionController::tryProcessingRequests()
             continue;
         }
 
+        auto* webPage = WebProcess::singleton().webPage(currentRequest.identifier);
```
Comment (cdumez) [inline Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp:102]: I think Sihui recommended this but I disagree with the suggestion:
1. The request should store the WebPageProxyIdentifier, not the PageIdentifier
2. There is no point in looking up the WebPage here if you already have the right identifier you need to send over IPC
3. This adds more complexity to try and avoid an IPC round trip in a very edge case where the WebPage doesn't exist. This is just not worth it.
4. Added more WebPage-related code is not a good idea because we're going to support service workers, which won't have a WebPage.

File: Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.h [hunk 18]
```diff
@@ -61,6 +59,7 @@ class WebPermissionController final : public CanMakeWeakPtr<WebPermissionControl
     struct PermissionRequest {
         WebCore::ClientOrigin origin;
         WebCore::PermissionDescriptor descriptor;
+        WebCore::PageIdentifier identifier;
```
Comment (cdumez) [inline Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.h:62]: Why a PageIdentifier when we need a WebPageProxyIdentifier?

**Review by cdumez — CHANGES_REQUESTED** [review]

File: Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp [hunk 19]
```diff
@@ -103,8 +99,14 @@ void WebPermissionController::tryProcessingRequests()
...
+        if (!webPage) {
```
Comment (szewai) [inline Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp:103]: This check should go before `queryCache`

File: Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp [hunk 20]
```diff
@@ -103,8 +99,14 @@ void WebPermissionController::tryProcessingRequests()
...
+            auto takenRequest = m_requests.takeFirst();
+            takenRequest.completionHandler(WebCore::PermissionState::Prompt);
```
Comment (szewai) [inline Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp:105]: Can be written in one line: `m_requests.takeFirst().completionHandler(cachedResult);`

File: Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp [hunk 21]
```diff
@@ -103,8 +99,14 @@ void WebPermissionController::tryProcessingRequests()
...
+        }
```
Comment (szewai) [inline Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp:106]: You need to call `continue` after completionHandler, otherwise you are going to send the message for this request.

File: Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp (hunk 17 above)
Comment (szewai) [reply Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp:102]: I don't think this is an edge case. In current impl, we block request handling when there's a pending request (waiting for UI process's reply). If a page makes multiple query requests and is gone, it's going to block requests from other pages for multiple rounds of IPCs.
I actually think we should store `std::optional<WeakPtr<WebPage>>` in the request, but PageIdentifier or `WeakPtr<WebPage>` might be enough for this patch.

**Review by szewai — CHANGES_REQUESTED** [review]

File: Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp (hunk 17 above)
Comment (cdumez) [reply Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp:102]: Very edge case. Multiple requests in parallel is an edge case, multiple pages in a single WebProcess is already an edge case. Even if there were really several and some queueing were to happen, querying doesn't involve prompting and we can thus expect responses really fast.

Also, an IPC round trip is cheap.

Comment (cdumez) [reply Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp:102]: I think you are adding complexity to try and optimize out a quick IPC round trip (quick since the UIProcess won't find the page and reply right away) for the case where:
1. There is more than one page in the WebProcess
and
2. Several of these WebPages are making query permission request in parallel
and
3. One of the WebPage gets closed before getting a response to its request(s)

File: Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp (hunk 16 above)
Comment (RupinMittal) [reply Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp:53]: We've gone back to using WebPageProxyIdentifier here. Fixed!

File: Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp (hunk 17 above)
Comment (RupinMittal) [reply Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp:102]: We've gone back to using WebPageProxyIdentifier here. Fixed!

File: Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp (hunk 20 above)
Comment (RupinMittal) [reply Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp:105]: We've gone back to using WebPageProxyIdentifier here and not checking the existence of the page.

File: Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp (hunk 21 above)
Comment (RupinMittal) [reply Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp:106]: We've gone back to using WebPageProxyIdentifier here and not checking the existence of the page.

File: Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.h (hunk 18 above)
Comment (RupinMittal) [reply Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.h:62]: We've gone back to using WebPageProxyIdentifier here--fixed!

**Review by cdumez — APPROVED** [review]
LGTM

**Review by szewai — APPROVED** [review]
Seems ready to land if bots are happy

File: Source/WebKit/WebKit.xcodeproj/project.pbxproj [hunk 22]
```diff
@@ -6717,6 +6720,12 @@
...
 		D3B9484511FF4B6500032B39 /* WebSearchPopupMenu.h */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.c.h; path = WebSearchPopupMenu.h; sourceTree = "<group>"; };
+		D952EC76289C4299006C240A /* WebPermissionControllerProxy.h */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.c.h; path = WebPermissionControllerProxy.h; sourceTree = "<group>"; };
+		D952EC77289C46B4006C240A /* WebPermissionControllerProxy.cpp */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.cpp.cpp; path = WebPermissionControllerProxy.cpp; sourceTree = "<group>"; };
+		D952EC79289C4AD0006C240A /* WebPermissionControllerProxy.messages.in */ = {isa = PBXFileReference; lastKnownFileType = text; path = WebPermissionControllerProxy.messages.in; sourceTree = "<group>"; };
+		D952EC7B289C9A4C006C240A /* WebPermissionControllerProxyMessages.h */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.c.h; name = WebPermissionControllerProxyMessages.h; path = ../Debug/DerivedSources/WebKit/WebPermissionControllerProxyMessages.h; sourceTree = BUILT_PRODUCTS_DIR; };
```
Comment (smfr) [inline Source/WebKit/WebKit.xcodeproj/project.pbxproj:6726]: Bad paths here.
